File: graph/ruby/adjlist_w.rb
class Graph
    def initialize(nv, edges)
        @nvertices = nv
        @adjlist = nv.times.collect { [] }
        edges.each do |i, j, w|
            @adjlist[i] << [j, w]
        end
    end
    def each_vertices
        @nvertices.times do |i|
            yield i
        end
    end
    def each_edges
        @nvertices.times do |i|
            @adjlist[i].each do |j, w|
                yield [i, j, w]
            end
        end
    end
    def each_successors(i)
        @adjlist[i].each do |j, w|
            yield [j, w]
        end
    end
    def vertices
        Enumerator.new(self, :each_vertices)
    end
    def edges
        Enumerator.new(self, :each_edges)
    end
    def successors(i)
        Enumerator.new(self, :each_successors, i)
    end
end

File: graph/haskell/AdjlistW.hs
module AdjlistW (
    Graph, Vertex, Edge,
    create, nvertices, vertices, edges, successors
    ) where

import Data.Array

type Vertex = Int
type Edge w = (Vertex, Vertex, w)
type Graph w = Array Vertex [(Vertex, w)]

create :: Num w => Int -> [Edge w] -> Graph w
create nv edges = accumArray (\vs v -> v:vs) []
    (0, nv-1) [(i,(j,w)) | (i,j,w) <- reverse edges]

nvertices :: Num w => Graph w -> Int
nvertices g = n + 1
    where
    (_, n) = bounds g

vertices :: Num w => Graph w -> [Vertex]
vertices g = [0..nvertices g]

edges :: Num w => Graph w -> [Edge w]
edges g = [(i,j,w) | (i, vs) <- assocs g, (j,w) <- vs]

successors :: Num w => Graph w -> Vertex -> [(Vertex,w)]
successors g i = g!i


